spinnaker-monitoring-daemon/spinnaker-monitoring/metric_filter.py

```diff
@@ -0,0 +1,201 @@
+
+    tags = spec.get('tags', {})
+    self.__include_tag_names_regex = self.__make_regex_list(
+        tags.get('includeNameRegex', []))
+    self.__exclude_tag_names_regex = self.__make_regex_list(
+        tags.get('excludeNameRegex', []))
+
+    if self.__exclude_tag_names_regex or self.__include_tag_names_regex:
+      self.__tag_filter = self.default_tag_filter
+    else:
+      self.__tag_filter = None
+
+  def __make_explicit_tag_func(self, spec):
+    """Create the tag function for explicitly specified metrics."""
+    class TagFunc(object):
+      def __init__(self, include_tags, exclude_tags):
+        self.__include_tags = include_tags
+        self.__exclude_tags = exclude_tags
+
+      def __call__(self, tag):
+        if self.__include_tags:
+          return tag in self.__include_tags
+        if self.__exclude_tags:
+          return tag not in self.__exclude_tags
+        return True
+
+    result = {}
+    for key, detail in spec.items():
+      if not detail:
+        result[key] = lambda x: True
+        continue
+      include_tags = set(detail.get('includeTags', []))
+      exclude_tags = set(detail.get('excludeTags', []))
+      if include_tags and exclude_tags:
+        raise ValueError(
+            'Cannot both include and exclude tags for metric "%s"' % key)
+      result[key] = TagFunc(include_tags, exclude_tags)
+
+    return result
+
+  def default_tag_filter(self, tag):
+    """For non-explicit metrics, determine if we should include tag."""
+    for regex in self.__exclude_tag_names_regex:
+      if regex.match(tag):
+        return False
+    for regex in self.__include_tag_names_regex:
+      if regex.match(tag):
+        return True
+    return True if self.__exclude_tag_names_regex else False
+
+  def combine_values(self, first, second):
+    """Combine the value lists by common timestamps."""
+    result = first
+    for record in second:
+      t = record['t']
+      v = record['v']
+      have = False
+      for check in result:
+        if check['t'] == t:
+          have = True
+          check['v'] += v
+          break
+      if not have:
+        result.append(record)
+        # Ensure values are in timestamp order
+        result = sorted(result, key=lambda x: x['t'])
+
+    return result
+
+  def filter_tags(self, data, tag_func):
+    """Mutate data to only keep tags of interest.
+
+    If tags are dropped, we'll need to aggregate the data.
+    For example if we had a values with:
+       {A='a', X='x'}=10
+       {A='a', X='y'}=5
+       {A='b', X='x'}=20
+    And we remove the tag X, then we have
+       {A='a'}=15
+       {A='b'}=20
+    """
+    metrics = data['values']
+    tags_hash_to_values = {}
+    hash_to_tags = {}
+    for metric in metrics:
+      values = metric['values']
+      tags = sorted([tag for tag in metric['tags']
+                     if tag_func(tag['key'])],
+                    key=lambda entry: entry['key'])
+      hash = str(tags)
```

Comment:
Might not matter too much, but `hash` is a python builtin: https://docs.python.org/2/library/functions.html.